component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.mutualtls/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/mutualtls/MutualTLSClientAuthenticator.java

```diff
@@ -174,16 +176,32 @@ public boolean authenticateClient(HttpServletRequest request, Map<String, List>
     public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams,
                                    OAuthClientAuthnContext context) {
 
-        if (clientIdExistsAsParam(bodyParams) && validCertExistsAsAttribute(request)) {
-            if (log.isDebugEnabled()) {
-                log.debug("Client ID exists in request body parameters and a valid certificate found in request " +
-                        "attributes. Hence returning true.");
+        String headerName = IdentityUtil.getProperty(CommonConstants.MTLS_AUTH_HEADER);
+        if (clientIdExistsAsParam(bodyParams)) {
+            if (validCertExistsAsAttribute(request)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("A valid certificate was found in the request attribute hence returning true.");
+                }
+                return true;
+
+            } else {
+                if (StringUtils.isNotBlank(headerName) && getCertificateFromHeader(request).isPresent()) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("A valid certificate was found from the request header hence returning true.");
+                    }
+                    return true;
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Mutual TLS authenticator cannot handle this request. " +
+                                "A valid certificate could not be found in the request.");
+                    }
+                    return false;
+                }
             }
-            return true;
-        } else {
+        }  else {
             if (log.isDebugEnabled()) {
-                log.debug("Mutual TLS authenticator cannot handle this request. Client id is not available in body " +
-                        "params or valid certificate not found in request attributes.");
+                log.debug("Mutual TLS authenticator cannot handle this request. " +
+                        "Client id is not available in body params.");
```

Comment:
Improve the log. eg:
"Client id is not available as a parameter in the body."

Reply:
Fixed 990e4d3